subprojects/docs/src/docs/userguide/groovy_plugin.adoc

```diff
@@ -253,6 +253,39 @@ include::sample[dir="groovy/compilationAvoidance/groovy", files="build.gradle[ta
 include::sample[dir="groovy/compilationAvoidance/kotlin", files="build.gradle.kts[tags=groovy-compilation-avoidance]"]
 ====
 
+[[sec:incremental_groovy_compile]]
+== Incremental Groovy compilation
+
+Since 5.6, Gradle introduces an experimental incremental Groovy compiler. To enable incremental compilation for Groovy, you need:
+
+* Enable <<groovy_plugin.adoc#sec:groovy_compilation_avoidance,Groovy compilation avoidance>>.
+* Explicitly enable incremental Groovy compilation in the build script:
+
+.Enable Groovy compilation
+====
+include::sample[dir="groovy/incrementalCompilation/groovy",files="build.gradle[tags=enable-groovy-incremental]"]
+include::sample[dir="groovy/incrementalCompilation/kotlin",files="build.gradle.kts[tags=enable-groovy-incremental]"]
+====
+
+This gives you the following benefits:
+
+* Incremental builds are much faster.
+* If only a small sets of Groovy source files are changed, only the affected source files will be recompiled.
+  Classes that don't need to be recompiled remain unchanged in the output directory.
+  For example, if you only change a few Groovy test classes, you don't need to recompile all Groovy test source files - only the changed ones need to be recompiled.
+
+To understand how incremental compilation works, see <<java_plugin.adoc#sec:incremental_compile,Incremental Java compilation>> for a detailed overview.
+Note that there're several differences from Java incremental compilation:
+
+* Unlike Java, Groovy compiler doesn't inline constants, thus changes to constants won't trigger a full recompilation.
+* Groovy compiler doesn't keep `@Retention` in generated bytecode, thus changes to annotations won't trigger a full recompilation.
```

Comment:
I am sure some annotations need to be kept in the generated bytecode, e.g. if the retention is runtime, right?

Reply:
I guess I need to clarify this. For an annotation class `MyAnnotation.groovy`, Groovy compiler doesn't keep `@Retention` in the `MyAnnotation.class` bytecode. See https://github.com/gradle/gradle/pull/9616#discussion_r294641010

Reply:
I guess you added this point because of this part in the incremental Java compilation documentation:
> Since source-retention annotations are not visible in bytecode, changes to a source-retention annotation will result in full recompilation.

I would leave it out, it adds more confusion than clarification. I don't also understand why the behaviour is specific to `@Retention`, and not about meta-annotations in general.